src/NLog/GlobalDiagnosticsContext.cs

```diff
@@ -158,21 +158,27 @@ public static void Clear()
             return readOnly;
         }
 
-        private static Dictionary<string, object> GetWritableDict(bool mustCreateNew, bool clone = true)
+        private static Dictionary<string, object> GetWritableDict(bool mustCreateNew, bool clearDictionary = false)
```

Comment:
This is a strange methode, with mustCreateNew and clear IMO. I tried to refactor that ;)

Reply:
It is very identical to MDC + MDLC - GetLogicalThreadDictionary

Reply:
well, clearDictionary  is only called once with `true`. Then it's strange to have that encapsulated in a method. It makes things a bit unnecessary complex.

Reply:
Yes I'm terrible at making code complex. But trying to improve, though I often forget. But I think https://github.com/snakefoot/NLog/pull/5 now is showing boiler-plate code for the different write-operations.

Reply:
Btw. the clearDictionary is an optimization-flag, so it knows it can skip the copying of the old dictionary.

Reply:
> showing boiler-plate code for the different write-operations.

true, set, remove and clear has code duplication, but that was already the case (in 1 line) and now those are 2 lines of duplication. Against it, CloneDictionary is much more straight-forward.

Reply:
Also GetWritableDict wasn't doing one thing, it was doing "get" but also "set" (but OK, that's also the case in MDC probably, but im trying to "fix the leak")

Reply:
With copy-on-write semantics then get-operation will sometimes be create/set. That is the nature of copy-on-write.

Reply:
from another view, do you think this line is easy to maintain?

`  GetWritableDict(_dictReadOnly != null && !_dict.ContainsKey(item))[item] = value;`

Reply:
> Yes I'm terrible good at making complex code.

I focus on having maintainable code, and unfortunately that's not always performant code :(  I think the best code for NLog is in-between :)

Reply:
Your points are very valid and valuable. Have now made another code-update for readability.